lib/elixir/src/elixir_erl_for.erl

```diff
@@ -135,24 +151,38 @@ build_into(Ann, Clauses, Expr, Into, Fun, Acc, S) ->
          elixir_erl:remote(Ann, erlang, raise, [Kind, Reason, Stack])]}],
       []},
 
-  {{block, Ann, [MatchExpr, TryExpr]}, SD}.
+  {{block, Ann, [MatchExpr, TryExpr]}, SI}.
 
 %% Helpers
 
-build_reduce(Clauses, Expr, false, Acc, S) ->
-  build_reduce_clause(Clauses, Expr, {nil, 0}, Acc, S);
-build_reduce(Clauses, Expr, {nil, Ann} = Into, Acc, S) ->
-  ListExpr = {cons, Ann, Expr, Acc},
-  elixir_erl:remote(Ann, lists, reverse,
-    [build_reduce_clause(Clauses, ListExpr, Into, Acc, S)]);
-build_reduce(Clauses, Expr, {bin, Ann, _} = Into, Acc, S) ->
-  BinExpr = {cons, Ann, Acc, Expr},
-  elixir_erl:remote(Ann, erlang, iolist_to_binary,
-    [build_reduce_clause(Clauses, BinExpr, Into, Acc, S)]).
-
-build_reduce_clause([{enum, Meta, Left, Right, Filters} | T], Expr, Arg, Acc, S) ->
+build_reduce(_Ann, Clauses, InnerFun, Expr, Into, false, Acc, S) ->
+  InnerExpr = InnerFun(Expr, Acc),
+  {build_reduce_each(Clauses, InnerExpr, Into, Acc, S), S};
+build_reduce(Ann, Clauses, InnerFun, Expr, Into, true, Acc, S) ->
```

Comment:
Actually I gave bad advice, we don't need to return `S` because all of the variables are used only inside the anonymous function. :) My bad, I will "revert" after merging!